Add TinyliciousAudience to the client and example of its consumption

sumedhb1995 opened the pull request:
The goals of this PR are the following:
- Create base interfaces for 
-- `IServiceAudience`, which will be implemented by each service-specific client to return audience information. This will provide the roster of current members with `getMembers`, and return the current user's information in `getMyself`. It also provides listeners for whenever the member list is changed and when a new edit is made.
--`IMember`, which can be extended by each service-specific client to provide details regarding member that can be unique based on the backing service
- Create a base class implementation of `IServiceAudience` with `ServiceAudience` in `fluid-static` that can be extended by each service-specific client package
- Extend the above for `ITinyliciousAudience` and `TinyliciousMember` and provide an implementation of `TinyliciousAudience`. 
- Show how we can use the audience object to keep track of our session members, and identify our current user is shown in the `external-controller` example. Here is a screenshot of the new view:
![image](https://user-images.githubusercontent.com/7992711/118885386-8d555e80-b8ac-11eb-9431-ba671242fda4.png)

Commit:
init implementation and example of consuming changes

Commit:
add copyright header

Commit:
include last edited functionality, support multiple clients per user, flatten user interface

Commit:
working integration with last-edited object

Commit:
wrap last edited in results object, comments

Commit:
update last edited naming

Commit:
drop public interface dep on protocol-def

Commit:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into tiny_audience_implementation

Commit:
use user names instead of ids for tiny

Commit:
comments and cleanup

Commit:
create base class ServiceAudience

Commit:
last edited tiny interface

Commit:
internally use last edited interface

ChumpChief commented on experimental/framework/tinylicious-client/src/TinyliciousClient.ts:
```diff
@@ -53,13 +56,13 @@ export class TinyliciousClientInstance {
         serviceContainerConfig: TinyliciousContainerConfig,
         containerSchema: ContainerSchema,
-    ): Promise<FluidContainer> {
+    ): Promise<[container: FluidContainer, containerServices: TinyliciousContainerServices]> {
```
Probably return an object here (and `createContainer`) so it's a little easier to change this in the future if we want.
```suggestion
    ): Promise<{container: FluidContainer, containerServices: TinyliciousContainerServices}> {
```

sumedhb1995 replied:
I converted this into a tuple since we had decided on the OdspClient that we would be returning a tuple containing the FluidContainer, which is common for all services, and the containerServices object, which would be unique for each one (https://github.com/microsoft/FluidM365Integration/pull/10#discussion_r621458318). The latter would be the one we could keep expanding on in the future to add in more capabilities

ChumpChief commented on experimental/framework/fluid-static/src/containerCode.ts:
```diff
@@ -87,11 +97,15 @@ export class RootDataObject
 
         await Promise.all(initialObjectsP);
+
+        const lastEditedFluidObject = await LastEditedTrackerDataObject.getFactory().createChildInstance(this.context);
+        this.root.set(this.lastEditedFluidObjectId, lastEditedFluidObject.handle);
```
We're currently trying to avoid built-in data objects in the core `ContainerRuntime` because (1) it's an unhelpful cost for customers who don't use them (e.g. if a customer doesn't care about LastEdited here, it's bundle size, op traffic, and storage use they don't benefit from) and also (2) it imposes a back-compat constraint on us to manage (e.g. suppose we come up with a better, but data-incompatible version of the last-edited tracker - making the change-over would break all existing customers).

Since the `FluidContainer` represents a hopeful future API surface for the `Container`, I think it would probably be best to keep this above the `FluidContainer` by forcing the caller to add it to the registry themselves, instantiate, etc. for similar reasons.  I think it could definitely go into a "framework" layer though.

sumedhb1995 replied:
Good point! I've removed the default inclusion of the last-edited-experimental object from the container. Instead, like you mentioned, developers can choose to opt-in to this feature, and perhaps we can include some tutorials to show how last-edited can be added to the initial containerSchema passed in and be fetched back by the developers themselves, instead of wrapping that functionality into the rootDataObject

ChumpChief commented on experimental/framework/fluid-static/src/types.ts:
```diff
@@ -71,3 +71,75 @@ export interface ContainerSchema {
+    /**
+     * Returns the member who made the last edit along with the edit's timestamp
+     */
+    getLastEdited(): ILastEditedResult<M> | undefined;
```
It seems difficult to scale/scope the duties of the audience if it also needs to contain accessory functionality like last-edited (e.g. as people want new information about the audience, do we have to keep expanding this interface?).

Is there some way we can have the last edited information live separately from the audience but be capable of correlating with audience members (e.g. through a unique ID or something)?

sumedhb1995 replied:
As we chatted about, I've removed the default consumption of the last-edited-experimental data object from the container. I think we can have that be separately added in by the user through the registry, and they could use the userId returned by the audience to separately correlate the values returned by last-edited to the members returned by the audience

ChumpChief commented on experimental/framework/tinylicious-client/src/TinyliciousAudience.ts:
```diff
@@ -0,0 +1,72 @@
+  /**
+   * @inheritdoc
+   */
+  public getMembers(): Map<string, TinyliciousMember> {
```
Is Map a good fit for usage scenarios here?  No suggestion to change at the moment, just thinking that it sort of assumes the user will want to retrieve by user ID.

Like the demo usage you added in external-controller looks like it would be just as happy with an array, for example.

sumedhb1995 replied:
Right, I was thinking that the user may have other places where they are storing values tagged with the user ID. If they ever need to correlate that with say the clients returned a specific user, they could do a simple look-up on this map. And if they want to use it as an array, they can just do `Array.from(getMembers().values())`.
I can convert it to an array but when we are preparing the response, we need to initially build a Map anyways to correlate multiple clients for the same user. We would just be moving the `Array.from` call internally instead of reducing the processing required

ChumpChief reviewed (commented):
Main feedback is to think about layering options to aid in scoping the responsibilities of the `RootDataObject` and the `IServiceAudience`.  I.e. so we can hopefully avoid the need to update the audience implementations for each new audience-related feature.

ChumpChief commented on experimental/framework/fluid-static/src/types.ts:
```diff
@@ -71,3 +71,75 @@ export interface ContainerSchema {
+    /**
+     * Returns the current active user on this client once they are connected. Otherwise, returns undefined.
+     */
+    getCurrentMember(): M | undefined;
```
I don't think "Current" implies the right thing here.  Maybe `getSelf()` or something?

sumedhb1995 replied:
Updated to be getMyself and getMyClient respectively!

ChumpChief commented on experimental/framework/fluid-static/src/serviceAudience.ts:
```diff
@@ -0,0 +1,167 @@
+    }
+  }
+
+  protected getMemberByClientId(clientId: string): IMember | undefined {
```
Looks like this could be private (at least for now, and can always be opened up more in the future)?

sumedhb1995 replied:
Converted to private

ChumpChief commented on experimental/framework/tinylicious-client/src/TinyliciousAudience.ts:
```diff
@@ -0,0 +1,72 @@
+import { IClient } from "@fluidframework/protocol-definitions";
+import { ITinyliciousAudience, TinyliciousLastEditedResult, TinyliciousMember } from "./interfaces";
+
+export class TinyliciousAudience extends ServiceAudience implements ITinyliciousAudience {
```
This is prob fine for now, but I'm always a little hesitant on class inheritance.  Since the service-specific audiences are generally made up of "decorated" members it seems maybe a good candidate to compose a `ServiceAudience` rather than inheriting from it.

sumedhb1995 replied:
ServiceAudience would work by default for any Fluid service since `IUser` always contains the user ID. Tinylicious would also provide the user name, and ODSP would further provide the user email.

The thinking here was that developers could extend the ServiceAudience class to include even more additional data to their Member objects beyond what is provided by Fluid. So, for example, if they wanted to add in more details to the IMember object that required making some API requests, they would have the ability to add that into their own implementation of `getMembers`.

Commit:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into tiny_audience_implementation

Commit:
self naming

Commit:
remove usage of last edited data object

Commit:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into tiny_audience_implementation

Commit:
update protocol-def version

sumedhb1995 commented:
@ChumpChief Went through the comments here and addressed them!
Along with some of the minor updates based on the comments, I've removed last-edited-object from the FluidContainer. The ServiceAudience still keeps track of the clients that have made edits after the user's own client joins the session, since this is just in-memory and does not require any additional data objects. As such, I've kept the listeners for whenever last edited changes. But we no longer keep track of the historical values from before the current client joins.

To properly support last-edited through providing it in the containerSchema, I think `last-edited-experimental` will need some updates to:
1) Not require a dependency on the container lifecycle and instead only on the component lifecycle, since the container lifecycle for `FluidContainer` should be generalized for any data objects added into the schema. `last-edited-experimental` currently needs the tracker to be fetched and setup in the `containerHasInitialized` function, like so:
```tsx
protected async containerHasInitialized(runtime: IContainerRuntime) {
        // Load the last edited tracker fluidObject (done by the setup method below). This fluidObject
        // provides container level tracking of last edit and has to be loaded before any other fluidObject.
        const tracker = await requestFluidObject<IFluidLastEditedTracker>(
            await runtime.getRootDataStore(rootDataStoreId),
            "",
        );
        setupLastEditedTrackerForContainer(tracker.IFluidLastEditedTracker, runtime);
    }
 ```
 2) Look into how some handshake can be setup to label ops by the DDS emitting them as something that is making a user-content change, and how `last-edited-experimental` can listen for those ops to update its state. This could be similar to the `IContainerRuntimeDirtyable` event that was deprecated, but whereas that was used to label **omit** AgentScheduler events, this would be used to specifically  **include** specific ops that DDS' and DataObjects could opt-in to

Commit:
restore last-edited changes

sumedhb1995 commented on experimental/framework/fluid-static/src/rootDataObject.ts:
```diff
@@ -27,6 +28,7 @@ interface RootDataObjectProps {
 
 // eslint-disable-next-line @typescript-eslint/ban-types
 export class RootDataObject extends DataObject<{}, RootDataObjectProps> {
+    private readonly opHandler = (message: ISequencedDocumentMessage) => this.emit("op", message);
```
opHandler is intentionally added here and not to FluidContainer as we don't want to expose this on the developer surface

ChumpChief commented on experimental/framework/fluid-static/src/types.ts:
```diff
@@ -71,3 +71,70 @@ export interface ContainerSchema {
+export interface IConnectedClient {
+    clientId: string;
+    connectionMode: "write" | "read";
+    timeLastActive?: Date;
```
I don't think we should internalize the last edited feature to the audience.  If the audience is a solid representation of connected clients then any metadata we want to track about those clients can be implemented as a standalone structure that observes the audience.  Internalizing metadata features invites future scope creep to the audience and also isn't needed by many customers.

For this PR I'd recommend just omitting the last edited functionality completely, and introduce it in a followup.

sumedhb1995 replied:
Last edited functionality has now been completely removed

ChumpChief commented on experimental/framework/fluid-static/src/types.ts:
```diff
@@ -71,3 +71,70 @@ export interface ContainerSchema {
+ * descending order from who made the most recent edit. This interface can be extended by each service
+ * to provide additional service-specific user metadata.
+ */
+export interface IMember {
```
I'm having a hard time keeping it straight what the difference/relationship between `IConnectedClient` vs `IMember` is (and similarly the API surface like "Myself" vs. "MyClient").  No change recommended for this PR necessarily, but wondering if there's some alternate naming we could consider to help clarify?

sumedhb1995 replied:
I believe with the change to remove any last edited state from audience, we can just keep the `getMyself` call and the `IMember` object. There shouldn't be anymore need for an `IConnectedClient` object. Also, perhaps in the future, we could name it "connection"? One member could have multiple `connectionId`s assigned to them based on connecting from different tabs or devices

sumedhb1995 replied:
I was thinking we could start calling them "connections" since it represents each unique connection to the session the member makes. I've renamed `IConnectedClient` here to be `IConnection` to avoid confusion also with multiple client-references, i.e. `TinyliciousClient`. Each `IMember` now holds a connections array.
`IConnection` now just holds the id (fetched from the container's clientId) and the mode (read vs write) to allow for any different UI representations based on which mode the client is joined in with, the last edited time is removed

ChumpChief replied:
Nice, I like that a lot better.

ChumpChief commented on experimental/framework/fluid-static/src/serviceAudience.ts:
```diff
@@ -0,0 +1,139 @@
+  /**
+   * @inheritdoc
+   */
+  public getMyClient(): IConnectedClient | undefined {
```
Should this exist here on the Audience?  I think the Container is the authority on clientId and connectionMode, Perhaps consulting with the (Fluid)Container directly for these items would be more straightforward.

sumedhb1995 replied:
Yep, if we're removing the timeLastActive from here, I don't really see any further reason to have this be exposed as an `IConnectedClient` object here. We can expose it directly from the `FluidContainer`

sumedhb1995 replied:
Removed! In a follow-up, perhaps we can restructure what is returned in FluidContainer to be an `IConnection` object holding the id and mode to allow for parity?

Commit:
add typedeventemitter, remove last edited client information as well

sumedhb1995 commented on experimental/framework/fluid-static/src/types.ts:
```diff
@@ -71,3 +71,70 @@ export interface ContainerSchema {
+    off(event: "membersChanged", listener: (members: M[]) => void): this;
+    off(event: "lastEditedChanged", listener: (result: ILastEditedResult<M>) => void): this;
+    once(event: "membersChanged", listener: (members: M[]) => void): this;
+    once(event: "lastEditedChanged", listener: (result: ILastEditedResult<M>) => void): this;
```
Added!

sumedhb1995 commented on experimental/framework/fluid-static/src/serviceAudience.ts:
```diff
@@ -0,0 +1,139 @@
+// Base class for providing audience information for sessions interacting with FluidContainer
+// This can be extended by different service-specific client packages to additional parameters to
+// the user and client details returned in IMember
+export class ServiceAudience extends EventEmitter implements IServiceAudience<IMember> {
```
Added! Also created the `IServiceAudienceEvents` interface to identify the `membersChanged` event

ChumpChief commented on experimental/framework/fluid-static/src/serviceAudience.ts:
```diff
@@ -0,0 +1,95 @@
+    }
+    // Return the member object with any other clients associated for this user
+    const allMembers = this.getMembers();
+    const lastEditedMember = allMembers.get(internalAudienceMember?.user.id);
+    if (lastEditedMember === undefined) {
+      throw Error(`Attempted to fetch client ${clientId} that is not part of the current member list`);
+    }
+    return lastEditedMember;
```
```suggestion
    const member = allMembers.get(internalAudienceMember?.user.id);
    if (member === undefined) {
      throw Error(`Attempted to fetch client ${clientId} that is not part of the current member list`);
    }
    return member;
```

ChumpChief commented on experimental/framework/fluid-static/src/types.ts:
```diff
@@ -71,3 +72,49 @@ export interface ContainerSchema {
+}
+
+/**
+ * Base interface to be implemented to fetch each service's audience. The generic C allows consumers to further
```
```suggestion
 * Base interface to be implemented to fetch each service's audience. The generic M allows consumers to further
```

ChumpChief commented on experimental/framework/tinylicious-client/src/TinyliciousAudience.ts:
```diff
@@ -0,0 +1,50 @@
+      // Get all the current human members
+      if (member.details.capabilities.interactive) {
+        const userId = member.user.id;
+        if (users.has(userId)) {
```
Similar to above

ChumpChief reviewed (approved):
Looks good!  A few comments/nits to consider but nothing blocking.

Commit:
address nits

sumedhb1995 commented on experimental/framework/fluid-static/src/serviceAudience.ts:
```diff
@@ -0,0 +1,95 @@
+      this.emit("membersChanged", this.getMembers());
+    });
+
+    this.container.on("connected", () => {
```
I had added the connected listener as the userId field for the current user (self) comes back as undefined until the user actually connects with the service. As such, we needed to refresh the list of members when the user connected.

For the disconnected case,
1) If it is another user who is disconnecting, it would trigger a "removeMember" event on the underlying container's audience and be emitted by `ServiceAudience` as a "membersChanged" event
2) If is the current user (self) who disconnected, they would know from the `FluidContainer`'s disconnected event

As such, in both cases, it looks like the event would be appropriately alerted to the developer?
